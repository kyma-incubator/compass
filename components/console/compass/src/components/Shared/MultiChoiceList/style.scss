.multi-choice-list__list-element {
    list-style-type: none;
    display: flex;
    justify-content: space-between;
    vertical-align: middle;
    align-items: center;

    & > :first-child {
      margin-left: 3px;
    }
}
