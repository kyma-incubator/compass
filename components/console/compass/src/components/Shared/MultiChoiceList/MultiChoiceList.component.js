import React from 'react';
import PropTypes from 'prop-types';

import { Menu, Dropdown, Popover, Button } from 'fundamental-react';
import './style.scss';

MultiChoiceList.propTypes = {
  placeholder: PropTypes.string,
  updateItems: PropTypes.func.isRequired,
  currentlySelectedItems: PropTypes.array.isRequired,
  currentlyNonSelectedItems: PropTypes.array.isRequired,
  notSelectedMessage: PropTypes.string,
  noEntitiesAvailableMessage: PropTypes.string,
  displayPropertySelector: PropTypes.string,
};

MultiChoiceList.defaultProps = {
  placeholder: 'Choose items...',
  noEntitiesAvailableMessage: 'No more entries available',
};

export default function MultiChoiceList({
  placeholder,
  currentlySelectedItems,
  currentlyNonSelectedItems,
  updateItems,
  notSelectedMessage,
  noEntitiesAvailableMessage,
  displayPropertySelector,
}) {
  function getDisplayName(item) {
    return displayPropertySelector ? item[displayPropertySelector] : item;
  }

  function selectItem(item) {
    const newSelectedItems = [...currentlySelectedItems, item];
    const newNonSelectedItems = currentlyNonSelectedItems.filter(
      i => i !== item,
    );

    updateItems(newSelectedItems, newNonSelectedItems);
  }

  function unselectItem(item) {
    const newSelectedItems = currentlySelectedItems.filter(i => i !== item);
    const newNonSelectedItems = [...currentlyNonSelectedItems, item];

    updateItems(newSelectedItems, newNonSelectedItems);
  }

  function createSelectedEntitiesList() {
    if (!currentlySelectedItems.length) {
      return <p className="fd-has-font-style-italic">{notSelectedMessage}</p>;
    }

    return (
      <ul>
        {currentlySelectedItems.map(item => (
          <li
            className="multi-choice-list__list-element"
            key={getDisplayName(item)}
          >
            <span>{getDisplayName(item)}</span>
            <Button
              data-test-id={`unselect-button`}
              option="light"
              type="negative"
              onClick={() => unselectItem(item)}
              size="l"
              glyph="decline"
            />
          </li>
        ))}
      </ul>
    );
  }

  function createNonSelectedEntitiesDropdown() {
    if (!currentlyNonSelectedItems.length) {
      return (
        <p className="fd-has-font-style-italic">{noEntitiesAvailableMessage}</p>
      );
    }

    const nonChoosenItemsList = (
      <Menu>
        {currentlyNonSelectedItems.map(item => (
          <Menu.Item
            key={getDisplayName(item)}
            onClick={() => selectItem(item)}
          >
            <span data-test-id={`select-button`}>{getDisplayName(item)}</span>
          </Menu.Item>
        ))}
      </Menu>
    );

    return (
      <Dropdown fullwidth="true">
        <Popover
          body={nonChoosenItemsList}
          control={
            <Button
              className="fd-dropdown__control"
              glyph="navigation-down-arrow"
            >
              {placeholder}
            </Button>
          }
          widthSizingType="matchTarget"
        />
      </Dropdown>
    );
  }

  return (
    <section className="multi-choice-list-modal__body">
      {createSelectedEntitiesList()}
      {createNonSelectedEntitiesDropdown()}
    </section>
  );
}
