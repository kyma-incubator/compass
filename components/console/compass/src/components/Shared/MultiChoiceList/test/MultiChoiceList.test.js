import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import * as mockData from './mockData';
import MultiChoiceList from '../MultiChoiceList.component';
import { Menu } from 'fundamental-react';

describe('MultiChoiceList', () => {
  // for "Warning: componentWillMount has been renamed"
  console.warn = jest.fn();
  // catch Warning: `NaN` is an invalid value for the `left` css style property from Popper
  console.error = jest.fn();

  afterEach(() => {
    console.warn.mockReset();
    console.error.mockReset();
  });

  afterAll(() => {
    if (console.warn.mock.calls.length) {
      expect(console.warn.mock.calls[0][0]).toMatchSnapshot();
    }
    if (console.error.mock.calls.length) {
      expect(console.error.mock.calls[0][0]).toMatchSnapshot();
    }
  });

  it('Renders empty list with default caption props', () => {
    const component = renderer.create(
      <MultiChoiceList
        updateItems={() => {}}
        currentlySelectedItems={[]}
        currentlyNonSelectedItems={[]}
      />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Renders empty list with custom caption props', () => {
    const component = renderer.create(
      <MultiChoiceList
        updateItems={() => {}}
        currentlySelectedItems={[]}
        currentlyNonSelectedItems={[]}
        placeholder="placeholder value"
        notSelectedMessage="not assigned message"
        noEntitiesAvailableMessage="no entities available message"
      />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Renders two lists of simple items', () => {
    const component = mount(
      <MultiChoiceList
        updateItems={() => {}}
        currentlySelectedItems={mockData.simpleSelected}
        currentlyNonSelectedItems={mockData.simpleNonselected}
      />,
    );

    const selectedItemTexts = component
      .find('.multi-choice-list-modal__body > ul')
      .text();
    expect(selectedItemTexts).toMatchSnapshot();

    // expand dropdown
    const dropdownControl = component.findWhere(
      t => t.text() === 'Choose items...' && t.type() == 'button',
    );

    dropdownControl.simulate('click');

    const nonSelectedItemTexts = component
      .find(Menu)
      .find('li')
      .map(node => node.text());
    expect(nonSelectedItemTexts).toMatchSnapshot();
  });

  it('Renders two lists of object items', () => {
    const component = mount(
      <MultiChoiceList
        updateItems={() => {}}
        currentlySelectedItems={mockData.selectedObjects}
        currentlyNonSelectedItems={mockData.nonSelectedObjects}
        displayPropertySelector="name"
      />,
    );

    const selectedItemTexts = component
      .find('.multi-choice-list-modal__body > ul')
      .text();
    expect(selectedItemTexts).toMatchSnapshot();

    // expand dropdown
    const dropdownControl = component.findWhere(
      t => t.text() === 'Choose items...' && t.type() == 'button',
    );
    dropdownControl.simulate('click');

    const nonSelectedItemTexts = component
      .find(Menu)
      .find('li')
      .map(node => node.text());
    expect(nonSelectedItemTexts).toMatchSnapshot();
  });

  it('Calls updateItems when user clicks on selected items list', () => {
    const updateItems = jest.fn();

    const component = mount(
      <MultiChoiceList
        updateItems={updateItems}
        currentlySelectedItems={['a', 'b']}
        currentlyNonSelectedItems={[]}
      />,
    );

    const removeItemButton = component
      .find('.multi-choice-list__list-element')
      .filterWhere(n => n.text() === 'a')
      .find('button[data-test-id="unselect-button"]');

    removeItemButton.simulate('click');

    expect(updateItems).toHaveBeenCalledWith(['b'], ['a']);
  });

  it('Calls updateItems when user clicks on non selected items list', () => {
    const updateItems = jest.fn();

    const component = mount(
      <MultiChoiceList
        updateItems={updateItems}
        currentlySelectedItems={[]}
        currentlyNonSelectedItems={['b', 'a']}
      />,
    );

    // expand list
    const popoverControl = component.find('.fd-button.fd-dropdown__control');
    popoverControl.simulate('click');

    component.update();

    const addItemButton = component
      .find('span[data-test-id="select-button"]')
      .filterWhere(n => n.text() === 'a');

    addItemButton.simulate('click');

    expect(updateItems).toHaveBeenCalledWith(['a'], ['b']);
  });
});
