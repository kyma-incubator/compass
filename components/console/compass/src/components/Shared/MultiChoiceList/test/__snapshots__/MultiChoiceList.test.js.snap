// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiChoiceList Renders empty list with custom caption props 1`] = `
<section
  className="multi-choice-list-modal__body"
>
  <p
    className="fd-has-font-style-italic"
  >
    not assigned message
  </p>
  <p
    className="fd-has-font-style-italic"
  >
    no entities available message
  </p>
</section>
`;

exports[`MultiChoiceList Renders empty list with default caption props 1`] = `
<section
  className="multi-choice-list-modal__body"
>
  <p
    className="fd-has-font-style-italic"
  />
  <p
    className="fd-has-font-style-italic"
  >
    No more entries available
  </p>
</section>
`;

exports[`MultiChoiceList Renders two lists of object items 1`] = `"choosen 1choosen 2"`;

exports[`MultiChoiceList Renders two lists of object items 2`] = `
Array [
  "non-choosen 1",
]
`;

exports[`MultiChoiceList Renders two lists of simple items 1`] = `"choosen 1choosen 2choosen 3"`;

exports[`MultiChoiceList Renders two lists of simple items 2`] = `
Array [
  "non-choosen 1",
  "non-choosen 2",
]
`;
